gems/workflow/app/models/workflow/application_record.rb

```diff
@@ -1,5 +1,27 @@
 module Workflow
   class ApplicationRecord < ActiveRecord::Base
     self.abstract_class = true
+
+    before_create :set_external_identifier
+    attr_readonly :external_identifier
+
+    def to_param
+      external_identifier
+    end
+
+    private
+
+    def generate_external_identifier
+      "#{SecureRandom.hex(2)}-#{SecureRandom.hex(2)}"
+    end
+
+    def set_external_identifier
+      return if external_identifier.present?
+
+      loop do
+        self.external_identifier = generate_external_identifier
+        break unless self.class.exists?(external_identifier: external_identifier)
+      end
+    end
   end
 end
```

Comment:
good solution.  althoguh i think this forces all models in the gem to have this.   but i think that's right for now?

Reply:
Ok, i'm going to move it to just affect the instances, b/c the definition tables don't even have an external id column yet.

Reply:
Hmmm, i'm actually having trouble figuring out how to put it into a library and including the code in the class.

Reply:
do you have a suggestion on how to do this so not all the models in the gym have external identifier?

Reply:
oh, good pt.  we will need that at some point if we make an admin UI for managing definitions that uses a react + API but honestly I'm thinking of just doing a rails app...   anyway just move that logic to the right models.